### PR TITLE
Conservative time integrator

### PR DESCRIPTION
**Pull request by vladotomov**
Added RK2Avg time integrator and other small modifications.

TODO:
- [x] Resolve differences in the results of run 7
- [x] Add a section in `README.md` for the Gresho vortex problem

### CODE REVIEW

**Comment by tzanio** [comment]
I get different results on verification run 7.

Can you check if that's due to this PR and update the README.md?

**Review by tzanio — APPROVED** [review]

**Comment by vladotomov** [comment]
It's not due to this PR, I get the same numbers when running master and this branch. With PA , on my mac, I get the numbers from the README, but different ones with FA. I'm not sure why.

File: README.md [hunk 1]
```diff
@@ -255,12 +255,11 @@ partial assembly option (`-pa`).
...
 ```sh
-mpirun -np 8 laghos -p 1 -m data/square01_quad.mesh -rs 3 -tf 0.8 -no-vis -pa
-mpirun -np 8 laghos -p 1 -m data/cube01_hex.mesh -rs 2 -tf 0.6 -no-vis -pa
+mpirun -np 8 laghos -p 1 -m data/square01_quad.mesh -rs 3 -tf 0.8 -pa
+mpirun -np 8 laghos -p 1 -m data/cube01_hex.mesh -rs 2 -tf 0.6 -vis -pa
 ```
 
-The latter produces the following density plot (when run with the `-vis` instead
-of the `-no-vis` option)
+The latter produces the following density plot (notice the `-vis` option)
```
Comment (tzanio) [inline README.md:262]: I like the Gresho vortex problem! Can we add a section on it, maybe with one of the following images?

glvis_s01
glvis_s03
glvis_s02

**Comment by tzanio** [comment]
> It's not due to this PR, I get the same numbers when running master and this branch. With PA , on my mac, I get the numbers from the README, but different ones with FA. I'm not sure why.

Here is what I get on my Linux box:

```
> mpirun -np 8 laghos -p 3 -m data/box01_hex.mesh -rs 1 -tf 3.0 -pa
...
step   925,     t = 3.0000,     dt = 0.001197,  |e| = 143.9954232321

CG (H1) total time: 44.3726212240
CG (H1) rate (megadofs x cg_iterations / second): 12.3904346832

CG (L2) total time: 0.7985124920
CG (L2) rate (megadofs x cg_iterations / second): 25.2604288625

Forces total time: 1.9640180770
Forces rate (megadofs x timesteps / second): 13.4134386585

UpdateQuadData total time: 12.9212457000
UpdateQuadData rate (megaquads x timesteps / second): 3.1462378275

Major kernels total time (seconds): 58.1460545530
Major kernels total rate (megadofs x time steps / second): 0.3660681737

Energy  diff: 2.95e-04
```

**Comment by tzanio** [comment]
Modulus the TODO items, I think this merge and v1.1 are ready to go.

Thanks @vladotomov!

**Comment by camierjs** [comment]
Both `master` and `time-integrators` give me the same final step/energy:
`step   948,     t = 3.0000,     dt = 0.002290,  |e| = 143.9965747577`
Removing the diagonal preconditioner  gives back
`step   937,     t = 3.0000,     dt = 0.002630,  |e| = 143.9989760246`
which is not yet exactly what is in the README, but much closer.

**Review by camierjs — APPROVED** [review]
Besides run 7 results that have to be updated, I'm good with this PR.

**Pull request merged**